### PR TITLE
Building MerkeTree Service

### PR DESCRIPTION
**Pull request by magestrio**
## Type

- [X] Feature
- [ ] Bug fix
- [ ] CI/CD
- [ ] Documentation
- [ ] Other

## Description

Brief description of the changes made.

### CODE REVIEW

File: packages/serverless/src/apollo/app/common.ts [hunk 1]
```diff
@@ -35,6 +35,6 @@ export const indexName = Joi.string()
   .required()
   .pattern(/^[_a-z]+[_a-z0-9]+/i);
 
-export const indexNumber = Joi.number().required();
+export const indexNumber = Joi.number().unsafe().required();
```
Comment (chiro-hiro) [inline packages/serverless/src/apollo/app/common.ts:38]: Incase of big number it should be encoded as a string and check with `/^[0-9]+/`.

File: packages/serverless/src/apollo/app/merkle-tree.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  sibling: String!
+}
+
+extend type Query {
+  getNode(databaseName: String!, level: Int!, index: BigInt!): String!
+  getWitness(databaseName: String!, root: String!, index: BigInt!): [MerkleProof]!
+}
+
+extend type Mutation {
+  createMerkleTree(databaseName: String!, height: Int!): Boolean
```
Comment (chiro-hiro) [inline packages/serverless/src/apollo/app/merkle-tree.ts:82]: I think we shouldn't allow users to manipulate the Merkle tree themselves. Instead we all users to perform data manipulation then we handle Merkle tree manipulation for them.

File: packages/serverless/src/helper/distributed-lock.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+import RedisInstance, { RedisClient } from './redis';
+import logger from './logger';
+
+const LOCK_KEY = 'myLockKey';
+const LOCK_EXPIRATION = 300; // in seconds
```
Comment (chiro-hiro) [inline packages/serverless/src/helper/distributed-lock.ts:5]: Is this blocking other instance?

**Comment by chiro-hiro** [comment]
I think we need to bind documents/collections with merkle tree, right now there is no bound.

File: packages/serverless/src/helper/distributed-lock.ts (hunk 3 above)
Comment (magestrio) [reply packages/serverless/src/helper/distributed-lock.ts:5]: Yes, only where the lock is applied

File: packages/serverless/src/apollo/app/merkle-tree.ts (hunk 2 above)
Comment (magestrio) [reply packages/serverless/src/apollo/app/merkle-tree.ts:82]: There are pretty many use cases when users need to work only with MT, aren't there?

File: packages/serverless/src/apollo/app/common.ts (hunk 1 above)
Comment (magestrio) [reply packages/serverless/src/apollo/app/common.ts]: Fixed

**Pull request merged**